clusterloader2/pkg/chaos/nodes.go

```diff
@@ -94,111 +51,20 @@ func (k *NodeKiller) Run(stopCh <-chan struct{}, wg *sync.WaitGroup) {
 	// wait.JitterUntil starts work immediately, so wait first.
 	sleepInterrupt(wait.Jitter(time.Duration(k.config.Interval), k.config.JitterFactor), stopCh)
 	wait.JitterUntil(func() {
-		nodes, err := k.pickNodes()
+		nodes, err := k.killProvider.PickNodes()
 		if err != nil {
 			klog.Errorf("%s: Unable to pick nodes to kill: %v", k, err)
 			return
 		}
-		k.kill(nodes, stopCh)
+		k.killProvider.Kill(nodes, stopCh)
 	}, time.Duration(k.config.Interval), k.config.JitterFactor, true, stopCh)
 }
 
-func (k *NodeKiller) pickNodes() ([]v1.Node, error) {
-	allNodes, err := util.GetSchedulableUntainedNodes(k.client)
-	if err != nil {
-		return nil, err
-	}
-
-	prometheusPods, err := client.ListPodsWithOptions(k.client, monitoringNamespace, metav1.ListOptions{
-		LabelSelector: prometheusLabel,
-	})
-	if err != nil {
-		return nil, err
-	}
-	nodesHasPrometheusPod := sets.NewString()
-	for i := range prometheusPods {
-		if prometheusPods[i].Spec.NodeName != "" {
-			nodesHasPrometheusPod.Insert(prometheusPods[i].Spec.NodeName)
-			klog.Infof("%s: Node %s removed from killing. Runs pod %s", k, prometheusPods[i].Spec.NodeName, prometheusPods[i].Name)
-		}
-	}
-
-	nodes := allNodes[:0]
-	for _, node := range allNodes {
-		if !nodesHasPrometheusPod.Has(node.Name) && !k.killedNodes.Has(node.Name) {
-			nodes = append(nodes, node)
-		}
-	}
-	rand.Shuffle(len(nodes), func(i, j int) {
-		nodes[i], nodes[j] = nodes[j], nodes[i]
-	})
-	numNodes := int(math.Ceil(k.config.FailureRate * float64(len(nodes))))
-	klog.Infof("%s: %d nodes available, wants to fail %d nodes", k, len(nodes), numNodes)
-	if len(nodes) > numNodes {
-		nodes = nodes[:numNodes]
-	}
-	for _, node := range nodes {
-		klog.Infof("%s: Node %q schedule for failure", k, node.Name)
-	}
-	return nodes, nil
-}
-
-func (k *NodeKiller) kill(nodes []v1.Node, stopCh <-chan struct{}) {
-	wg := sync.WaitGroup{}
-	wg.Add(len(nodes))
-	for _, node := range nodes {
-		k.killedNodes.Insert(node.Name)
-		node := node
-		go func() {
-			defer wg.Done()
-
-			klog.Infof("%s: Stopping docker and kubelet on %q to simulate failure", k, node.Name)
-			k.addStopServicesEvent(node.Name)
-			err := util.SSH("sudo systemctl stop docker kubelet", &node, nil)
```

Comment:
Here we should only extract `err := util.SSH("sudo systemctl stop docker kubelet", &node, nil)` and `err = util.SSH("nohup sudo shutdown -r +1s > /dev/null 2> /dev/null < /dev/null &", &node, nil)` as provider specific code.

Shared code can be unit tested, provider specific code needs integration testing and it's harder to do locally.